File: src/carmine/pigment.py
"""Pigment application in CIELAB space.

Painting a flat RGB color straight onto a masked region (or blending it in
additively) either erases skin texture or blows out brightness. Every op
here does its work in Lab: chroma (a/b) moves toward the target color while
lightness (L) mostly keeps the original per-pixel detail, so tinted skin
still looks like skin instead of a sticker. ``paint`` is the deliberate
exception -- it collapses the region to a flat color, which is exactly what
a hard-edged product like eyeliner needs.

The finish functions (``finish_matte``, ``finish_gloss``) reshape the L
channel's distribution within a mask to fake a product's surface behavior:
matte damps local highlights toward their blurred neighborhood, gloss lifts
the brightest highlights further above the rest.
"""

from __future__ import annotations

import re

import cv2
import numpy as np

_HEX_RE = re.compile(r"^#?([0-9a-fA-F]{6})$")


def parse_hex_color(value: str) -> tuple[int, int, int]:
    """Parse a ``#RRGGBB`` (or bare ``RRGGBB``) string into an (R, G, B) tuple.

    Raises:
        ValueError: `value` is not a string, or doesn't match the pattern.
        The offending value is embedded in the message.
    """
    if not isinstance(value, str):
        raise ValueError(f"color must be a '#RRGGBB' string, got {value!r}")
    match = _HEX_RE.match(value.strip())
    if not match:
        raise ValueError(f"invalid hex color {value!r}, expected '#RRGGBB'")
    hex_digits = match.group(1)
    return (
        int(hex_digits[0:2], 16),
        int(hex_digits[2:4], 16),
        int(hex_digits[4:6], 16),
    )


def _to_lab(image_bgr: np.ndarray) -> np.ndarray:
    return cv2.cvtColor(image_bgr.astype(np.float32) / 255.0, cv2.COLOR_BGR2Lab)


def _from_lab(lab: np.ndarray) -> np.ndarray:
    bgr = cv2.cvtColor(lab, cv2.COLOR_Lab2BGR)
    return np.clip(bgr * 255.0, 0, 255).astype(np.uint8)


def _color_to_lab(color_rgb: tuple[int, int, int]) -> np.ndarray:
    swatch = np.array([[color_rgb[::-1]]], dtype=np.float32) / 255.0  # RGB -> BGR
    return cv2.cvtColor(swatch, cv2.COLOR_BGR2Lab)[0, 0]


def _restore_untouched(out: np.ndarray, image_bgr: np.ndarray, mask: np.ndarray) -> np.ndarray:
    """Copy original pixels back wherever the mask is zero.

    The Lab round-trip (float conversion, color-space matrix multiply,
    conversion back to uint8) drifts pixels by a few sRGB levels even at
    weight 0, purely from floating-point rounding. Callers rely on
    unmasked pixels staying bit-identical to the input, so we paper over
    that drift explicitly rather than chase it out of the color math.
    """
    untouched = mask <= 0
    out[untouched] = image_bgr[untouched]
    return out


def tint(
    image_bgr: np.ndarray,
    mask: np.ndarray,
    color_rgb: tuple[int, int, int],
    intensity: float,
    lightness_pull: float = 0.35,
) -> np.ndarray:
    """Pull the masked region's chroma toward `color_rgb` in Lab space.

    `intensity` scales the whole effect. `lightness_pull` caps how much of
    the target's own lightness bleeds in -- kept well under 1 so shading
    and highlights already in the image survive the tint.
    """
    if intensity <= 0:
        return image_bgr.copy()
    lab = _to_lab(image_bgr)
    target = _color_to_lab(color_rgb)
    weight = (mask * float(intensity))[..., None]
    lightness_weight = weight * lightness_pull
    lab[..., 1:] += (target[1:] - lab[..., 1:]) * weight
    lab[..., :1] += (target[0] - lab[..., :1]) * lightness_weight
    out = _from_lab(lab)
    return _restore_untouched(out, image_bgr, mask)


def paint(
    image_bgr: np.ndarray,
    mask: np.ndarray,
    color_rgb: tuple[int, int, int],
    intensity: float,
) -> np.ndarray:
    """Alpha-blend a flat pigment over the masked region.

    Unlike `tint`, this discards the underlying texture entirely -- the
    right behavior for a hard-edged product like eyeliner, where covering
    what's underneath is the point rather than a side effect.
    """
    if intensity <= 0:
        return image_bgr.copy()
    color = np.array(color_rgb[::-1], dtype=np.float32)
    weight = (mask * float(intensity))[..., None]
    out = image_bgr.astype(np.float32) * (1 - weight) + color * weight
    return np.clip(out, 0, 255).astype(np.uint8)


def smooth(image_bgr: np.ndarray, mask: np.ndarray, amount: float) -> np.ndarray:
    """Blend in an edge-preserving bilateral filter within the masked region."""
    if amount <= 0:
        return image_bgr.copy()
    softened = cv2.bilateralFilter(image_bgr, d=9, sigmaColor=45, sigmaSpace=9)
    weight = (mask * float(amount))[..., None]
    out = image_bgr.astype(np.float32) * (1 - weight) + softened.astype(np.float32) * weight
    return np.clip(out, 0, 255).astype(np.uint8)


def finish_matte(image_bgr: np.ndarray, mask: np.ndarray, strength: float = 0.35) -> np.ndarray:
    """Flatten micro-highlights within the mask toward their local blur.

    Pulls each masked pixel's L value toward a sigma=5 Gaussian blur of L,
    which damps small specular highlights while leaving the broad shading
    that reads as skin/lip shape intact.
    """
    if strength <= 0:
        return image_bgr.copy()
    lab = _to_lab(image_bgr)
    l_channel = lab[..., 0]
    blurred = cv2.GaussianBlur(l_channel, (0, 0), sigmaX=5)
    weight = mask * float(strength)
    lab[..., 0] = l_channel + (blurred - l_channel) * weight
    out = _from_lab(lab)
    return _restore_untouched(out, image_bgr, mask)


def finish_gloss(image_bgr: np.ndarray, mask: np.ndarray, strength: float) -> np.ndarray:
    """Boost the brightest highlights within the mask to fake specular shine.

    Builds a highlight map from where each masked pixel's L falls between
    the 75th and 99th percentile of L within the mask, then adds to L in
    proportion to that map, `mask`, and `strength`. Regions too small or
    too flat to have a meaningful highlight spread are left untouched.
    """
    if strength <= 0:
        return image_bgr.copy()
    lab = _to_lab(image_bgr)
    l_channel = lab[..., 0]
    inside = mask > 0.5
    if int(np.count_nonzero(inside)) < 10:
        return image_bgr.copy()
    masked_l = l_channel[inside]
    p75, p99 = np.percentile(masked_l, [75, 99])
    spread = p99 - p75
    if spread < 1e-6:
        return image_bgr.copy()
    highlight = np.clip((l_channel - p75) / spread, 0.0, 1.0)
    lab[..., 0] = np.clip(
        l_channel + highlight * mask * 18.0 * float(strength), 0, 100
    )
    out = _from_lab(lab)
    return _restore_untouched(out, image_bgr, mask)

File: web/src/engine/pigment.ts
/**
 * CPU pigment ops in CIELAB -- a port of `carmine/pigment.py`.
 *
 * Painting a flat RGB color into a masked region either erases skin texture
 * or blows out brightness, so every operation here works in Lab: chroma
 * (a/b) moves toward the target color while lightness (L) keeps most of its
 * per-pixel detail, and tinted skin still reads as skin. `paint` is the
 * deliberate exception -- it collapses the region to a flat color, which is
 * what a hard-edged product like eyeliner needs.
 *
 * This module is the CPU reference path. The live camera path runs the same
 * math in a fragment shader (`renderer.ts`); this implementation is what the
 * parity tests measure against the Python engine, and what the gloss
 * percentile helper feeds the shader from.
 *
 * **Pixel format.** Images are flat RGBA `Float32Array`s at processing
 * resolution holding *byte-valued* floats (0-255 integers), which is what
 * `ImageData` gives us once widened. Outputs are clamped to [0, 255] and
 * **truncated**, not rounded, because NumPy's `.astype(np.uint8)` truncates
 * -- rounding instead would put roughly half the pixels one level away from
 * the Python result.
 */

import { labToRgb, rgbToLab, type Rgb } from "./color";
import { gaussianBlur } from "./blur";
import constants from "../gen/constants.json";

const MATTE_BLUR_SIGMA = constants.pigment.finish_matte.blur_sigma;
const GLOSS_P_LOW = constants.pigment.finish_gloss.percentile_low;
const GLOSS_P_HIGH = constants.pigment.finish_gloss.percentile_high;
const GLOSS_STRENGTH_FACTOR = constants.pigment.finish_gloss.strength_factor;

/** Minimum masked-pixel count below which gloss has nothing meaningful to
 * measure a highlight spread from (pigment.py finish_gloss). */
const GLOSS_MIN_PIXELS = 10;

/** Percentile pair describing where a mask's highlights sit on the L axis. */
export interface GlossPercentiles {
  p75: number;
  p99: number;
}

function toByte(value: number): number {
  // clamp then truncate: NumPy's np.clip(x, 0, 255).astype(np.uint8).
  if (value <= 0) return 0;
  if (value >= 255) return 255;
  return Math.trunc(value);
}

/**
 * Copy original pixels back wherever the mask is zero.
 *
 * The Lab round trip drifts pixels by a level or two even at weight 0,
 * purely from floating-point rounding. Callers rely on untouched regions
 * staying bit-identical to the input -- a look with one product enabled
 * must not perturb the rest of the frame -- so the drift is papered over
 * explicitly here rather than chased out of the color math.
 */
function restoreUntouched(out: Float32Array, source: Float32Array, mask: Float32Array): void {
  for (let i = 0; i < mask.length; i++) {
    if (mask[i] <= 0) {
      const p = i * 4;
      out[p] = source[p];
      out[p + 1] = source[p + 1];
      out[p + 2] = source[p + 2];
      out[p + 3] = source[p + 3];
    }
  }
}

/**
 * Pull the masked region's chroma toward `colorRgb` in Lab space.
 *
 * `intensity` scales the whole effect; `lightnessPull` caps how much of the
 * target's own lightness bleeds in, kept well under 1 so the shading and
 * highlights already in the frame survive the tint.
 */
export function tint(
  pixels: Float32Array,
  mask: Float32Array,
  colorRgb: Rgb,
  intensity: number,
  lightnessPull: number,
): Float32Array {
  const out = pixels.slice();
  if (intensity <= 0) {
    return out;
  }
  const target = rgbToLab(colorRgb);

  for (let i = 0; i < mask.length; i++) {
    const weight = mask[i] * intensity;
    if (weight === 0) {
      continue;
    }
    const p = i * 4;
    const lab = rgbToLab([pixels[p], pixels[p + 1], pixels[p + 2]]);
    lab[1] += (target[1] - lab[1]) * weight;
    lab[2] += (target[2] - lab[2]) * weight;
    lab[0] += (target[0] - lab[0]) * weight * lightnessPull;
    const rgb = labToRgb(lab);
    out[p] = toByte(rgb[0]);
    out[p + 1] = toByte(rgb[1]);
    out[p + 2] = toByte(rgb[2]);
  }

  restoreUntouched(out, pixels, mask);
  return out;
}

/**
 * Alpha-blend a flat pigment over the masked region.
 *
 * Unlike `tint` this discards the underlying texture entirely, which is the
 * right behavior for eyeliner: covering what is beneath is the point rather
 * than a side effect. No Lab round trip is involved, so no untouched-restore
 * pass is needed -- a zero weight already leaves the pixel exact.
 */
export function paint(
  pixels: Float32Array,
  mask: Float32Array,
  colorRgb: Rgb,
  intensity: number,
): Float32Array {
  const out = pixels.slice();
  if (intensity <= 0) {
    return out;
  }
  for (let i = 0; i < mask.length; i++) {
    const weight = mask[i] * intensity;
    if (weight === 0) {
      continue;
    }
    const p = i * 4;
    for (let c = 0; c < 3; c++) {
      out[p + c] = toByte(pixels[p + c] * (1 - weight) + colorRgb[c] * weight);
    }
  }
  return out;
}

/** Extract the L channel of an RGBA image as a single-channel Float32Array. */
function lightnessChannel(pixels: Float32Array, count: number): Float32Array {
  const l = new Float32Array(count);
  for (let i = 0; i < count; i++) {
    const p = i * 4;
    l[i] = rgbToLab([pixels[p], pixels[p + 1], pixels[p + 2]])[0];
  }
  return l;
}

/** Rewrite an RGBA image's pixels from a modified L channel, keeping each
 * pixel's original a/b. */
function applyLightness(
  pixels: Float32Array,
  lightness: Float32Array,
  out: Float32Array,
  changed: (index: number) => boolean,
): void {
  for (let i = 0; i < lightness.length; i++) {
    if (!changed(i)) {
      continue;
    }
    const p = i * 4;
    const lab = rgbToLab([pixels[p], pixels[p + 1], pixels[p + 2]]);
    lab[0] = lightness[i];
    const rgb = labToRgb(lab);
    out[p] = toByte(rgb[0]);
    out[p + 1] = toByte(rgb[1]);
    out[p + 2] = toByte(rgb[2]);
  }
}

/**
 * Flatten micro-highlights within the mask toward their local blur.
 *
 * Pulls each masked pixel's L toward a sigma=5 Gaussian blur of the whole L
 * channel, damping small specular highlights while leaving the broad shading
 * that reads as lip shape intact. The blur is taken over the *entire*
 * channel, not just the mask, so the pull near a mask edge references real
 * neighboring skin rather than zeros.
 */
export function finishMatte(
  pixels: Float32Array,
  width: number,
  height: number,
  mask: Float32Array,
  strength: number,
): Float32Array {
  const out = pixels.slice();
  if (strength <= 0) {
    return out;
  }
  const lightness = lightnessChannel(pixels, width * height);
  const blurred = gaussianBlur(lightness, width, height, MATTE_BLUR_SIGMA);
  const adjusted = new Float32Array(lightness.length);
  for (let i = 0; i < lightness.length; i++) {
    adjusted[i] = lightness[i] + (blurred[i] - lightness[i]) * mask[i] * strength;
  }
  applyLightness(pixels, adjusted, out, () => true);
  restoreUntouched(out, pixels, mask);
  return out;
}

/** NumPy's default (`linear`) percentile over an already-sorted array. */
function percentileSorted(sorted: Float32Array, q: number): number {
  const position = (q / 100) * (sorted.length - 1);
  const lower = Math.floor(position);
  const upper = Math.ceil(position);
  if (lower === upper) {
    return sorted[lower];
  }
  return sorted[lower] + (sorted[upper] - sorted[lower]) * (position - lower);
}

/**
 * The 75th and 99th L percentiles inside `mask`, or null when the region is
 * too small or too flat for a highlight to mean anything.
 *
 * This is split out from `finishGloss` because the WebGL path needs the same
 * two numbers as uniforms: percentiles are a whole-region reduction, which a
 * fragment shader cannot do, so they are computed once per frame on the CPU
 * from the downscaled masked image and handed to the shader.
 */
export function glossPercentiles(
  pixels: Float32Array,
  mask: Float32Array,
): GlossPercentiles | null {
  const values: number[] = [];
  for (let i = 0; i < mask.length; i++) {
    if (mask[i] > 0.5) {
      const p = i * 4;
      values.push(rgbToLab([pixels[p], pixels[p + 1], pixels[p + 2]])[0]);
    }
  }
  if (values.length < GLOSS_MIN_PIXELS) {
    return null;
  }
  const sorted = Float32Array.from(values).sort();
  const p75 = percentileSorted(sorted, GLOSS_P_LOW);
  const p99 = percentileSorted(sorted, GLOSS_P_HIGH);
  if (p99 - p75 < 1e-6) {
    return null;
  }
  return { p75, p99 };
}

/**
 * Boost the brightest highlights within the mask to fake specular shine.
 *
 * Builds a highlight map from where each masked pixel's L falls between the
 * 75th and 99th percentile of L inside the mask, then adds to L in
 * proportion to that map, the mask, and `strength`.
 */
export function finishGloss(
  pixels: Float32Array,
  mask: Float32Array,
  strength: number,
  percentiles?: GlossPercentiles | null,
): Float32Array {
  const out = pixels.slice();
  if (strength <= 0) {
    return out;
  }
  const stats = percentiles === undefined ? glossPercentiles(pixels, mask) : percentiles;
  if (stats === null) {
    return out;
  }
  const spread = stats.p99 - stats.p75;
  const count = mask.length;
  const lightness = lightnessChannel(pixels, count);
  const adjusted = new Float32Array(count);
  for (let i = 0; i < count; i++) {
    const highlight = Math.min(Math.max((lightness[i] - stats.p75) / spread, 0), 1);
    const boosted = lightness[i] + highlight * mask[i] * GLOSS_STRENGTH_FACTOR * strength;
    adjusted[i] = Math.min(Math.max(boosted, 0), 100);
  }
  applyLightness(pixels, adjusted, out, () => true);
  restoreUntouched(out, pixels, mask);
  return out;
}
